sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -1071,3 +1072,29 @@ trait UnresolvedPlanId extends LeafExpression with Unevaluable {
   // Subclasses can override this function to provide more TreePatterns.
   def nodePatternsInternal(): Seq[TreePattern] = Seq()
 }
+
+case class UnresolvedWithColumns(
+     colNames: Seq[String],
+     exprs: Seq[Expression],
+     metadata: Option[Seq[Metadata]],
+     child: LogicalPlan)
+  extends UnresolvedUnaryNode {
+
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_WITH_COLUMNS)
+
+  override protected def withNewChildInternal(
+      newChild: LogicalPlan): UnresolvedWithColumns = copy(child = newChild)
+}
+
+case class UnresolvedWithColumnsRenamed(
```

Comment:
I don't think we need new logical plans. See how we implement a similar feature for SQL pipe: https://github.com/apache/spark/commit/68be1da5ba8e54a67eb61822cdf7f8eb47926319

Reply:
We can reuse `Project` with `UnresolvedStarExceptOrReplace`

Reply:
oh, `df.withColumn` can append or replace columns, we should probably extend `UnresolvedStarExceptOrReplace` for this case.

Reply:
cc @dtenedor

Reply:
There seem to be some semantic differences even with `withColumnsRenamed`:

- If the specified column names are missing in the `df`, `df.withColumnsRenamed` ignores whereas `UnresolvedStarExceptOrReplace` throws an exception.
- `df.withColumnsRenamed` respects the argument order, e.g., 
    ```
    test("SPARK-46260: withColumnsRenamed should respect the Map ordering") {
      val df = spark.range(10).toDF()
      assert(df.withColumnsRenamed(ListMap("id" -> "a", "a" -> "b")).columns === Array("b"))
      assert(df.withColumnsRenamed(ListMap("a" -> "b", "id" -> "a")).columns === Array("a"))
    }
    ```
    whereas `UnresolvedStarExceptOrReplace` throws an exception.

Reply:
I see. One question is shall we add new logical plans or new star-like expressions? star-like expressions are more flexible as they work for function input as well (e.g. `struct(*)`). And we might be able to unify them with the SQL pipe one with extra flags.

Reply:
Moved to the star-like expressions.